downloader/src/main/kotlin/VcsHost.kt

```diff
@@ -0,0 +1,172 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package com.here.ort.downloader
+
+import com.here.ort.model.VcsInfo
+import com.here.ort.model.VcsType
+import com.here.ort.utils.hasFragmentRevision
+
+import java.net.URI
+import java.net.URISyntaxException
+import java.nio.file.Paths
+
+import kotlin.reflect.full.primaryConstructor
+
+/**
+ * A class to handle VCS-host-specific information.
+ */
+sealed class VcsHost(val url: String) {
+    companion object {
+        private val ALL = VcsHost::class.sealedSubclasses
+
+        /**
+         * Get the VCS host that is applicable for [url].
+         */
+        fun fromUrl(url: String): VcsHost? =
+            ALL.asSequence().map {
+                it.primaryConstructor!!.call(url)
+            }.find {
+                it.isApplicable()
+            }
+    }
+
+    /**
+     * The [URI] for [url] if it can be parsed, or null otherwise.
+     */
+    protected val uri = try {
+        URI(url)
+    } catch (e: URISyntaxException) {
+        null
+    }
+
+    /**
+     * The base hostname of VCS host.
+     */
+    abstract val hostname: String
+
+    /**
+     * Whether the VCS host can only host Git repositories.
+     */
+    abstract val isGitOnly: Boolean
+
+    /**
+     * Return whether this instance is applicable for the [uri] it was constructed with.
+     */
+    fun isApplicable() = uri?.host == hostname
+
+    /**
+     * Return all [VcsInfo] that can be extracted from [uri].
+     */
+    abstract fun toVcsInfo(): VcsInfo?
+}
+
+/**
+ * A class to handle [Bitbucket][https://bitbucket.org/]-specific information.
+ */
+class Bitbucket(url: String) : VcsHost(url) {
+    override val hostname = "bitbucket.org"
+    override val isGitOnly = false
+
+    override fun toVcsInfo(): VcsInfo? {
+        if (uri == null) return null
+        var url = uri.scheme + "://" + uri.authority
+
+        // Append the first two path components that denote the user and project to the base URL.
+        val pathIterator = Paths.get(uri.path).iterator()
+        if (pathIterator.hasNext()) {
+            url += "/${pathIterator.next()}"
+        }
+        if (pathIterator.hasNext()) {
+            url += "/${pathIterator.next()}"
+        }
+
+        var revision = ""
+        var path = ""
+
+        if (pathIterator.hasNext() && pathIterator.next().toString() == "src") {
+            if (pathIterator.hasNext()) {
+                revision = pathIterator.next().toString()
+                path = uri.path.substringAfter(revision).trimStart('/').removeSuffix(".git")
+            }
+        }
+
+        val type = VersionControlSystem.forUrl(url)?.type ?: VcsType.NONE
+        if (type == VcsType.GIT) {
+            url += ".git"
+        }
+
+        return VcsInfo(type, url, revision, path = path)
+    }
+}
+
+private fun gitUrlToVcsInfo(uri: URI): VcsInfo {
```

Comment:
I would prefer to have this function above Bitbucket or below GitLab instead, to not have it in the middle of the subclasses, but approved anyway.

Reply:
Or how about even above the `VcsHost` class?

Reply:
I'm not a fan of not starting a file with the class of the same name, but I would prefer that over having it in between the subclasses.

Reply:
Ok, as to me it doesn't really make a difference, I'll move the function to above Bitbucket, as you preferred.